### ISSUE TITLE
[Utility] Per-chain Relays To Token Multiplier

### ISSUE THREAD

**Issue by Olshansk** [pokt-network/poktroll#494]
## Objective

Feature parity with RTTM from Morse.

## Origin Document

- #437
- [Pocket Forum PIP33](https://forum.pokt.network/t/pip33-unleashing-the-potential-of-pocket-as-universal-rpc-provider/4797/8)
- https://github.com/pokt-network/pocket-core/pull/1580

## Goals

- Achieve RTTM parity w/ Morse

## Deliverables

- [ ] **Research Morse**: Study & understand how RTTM works in Morse
- [ ] **Governance**: Add the RTTM governance parameter
- [ ] **Implement**: Implement RTTM in Shannon
- [ ] **Integrate**: Make sure the RTTM values are probability capture in the Sparse Merkle Trie
- [ ] **CLI**:  Update the `poktrolld` (if necessary) to support updating the RTTM governance parameter
- [ ] **Test**: Add unit, integration and E2E tests
- [ ] **Tooling**: Add `Makefile` targets to easily trigger (on LocalNet) and test (e2e) modifying / updating / testing different RTTM values
- [ ] **Document**: Add a new page to dev.poktroll.com explaining how RTTM works and how to use it
- [ ] **Comments**: Add/update TODOs and comments alongside the source code so it is easier to follow.

## Non-goals / Non-deliverables

- Expanding beyond what Morse has for RTTM (except integration into the SMT)

## Estimated Days of Work

**2 days**

_Disclaimer_: This is the total projected number of estimated hours to completion & merge. The owner of this tickets is expected to use this GitHub issue to communicate with the core protocol team along the way, with update & feedback for each deliverable throughout the duration of this work._

---

**Creator**: @Olshansk 
**Co-Owners**: @moatus

**Comment by rBurgett** [comment]
@Olshansk I'd be happy to take this.

**Comment by Olshansk** [comment]
@rBurgett Sounds great!

Few questions / suggestions (some redundancy from our offline conversations but posting here for posterity)

1. **Q1**: When do you think you'll be able to start?
2. **S1**: Please start with a draft PR and cc the team (@bryanchriswhite, @Olshansk, @red-0ne) before diving too deep into the implementation, tests, docs, etc...
3. **S2**: Make sure to look into `ignite` to see if there is any scaffolding opportunity (less work, standards, etc...)

CCing @moatus for visibility.

**Comment by rBurgett** [comment]
I'll start tomorrow and work on it as I can around my current, full work schedule. I'll have a draft PR up by Monday.

**Comment by Olshansk** [comment]
![Screenshot 2024-06-05 at 1 01 39 PM](https://github.com/pokt-network/poktroll/assets/1892194/61094af7-63b5-4af2-a33f-bfcd0d3a3b7b)
![Screenshot 2024-06-05 at 1 01 43 PM](https://github.com/pokt-network/poktroll/assets/1892194/a9a23cc8-a9f0-41b5-b06c-88881b3dceb8)
![Screenshot 2024-06-05 at 1 01 47 PM](https://github.com/pokt-network/poktroll/assets/1892194/dd8118b3-63e9-4b13-9990-18f2f3a8eaf4)
![Screenshot 2024-06-05 at 1 02 20 PM](https://github.com/pokt-network/poktroll/assets/1892194/7b58105e-11d7-47b2-98a3-29b354914f91)
![Screenshot 2024-06-05 at 1 02 28 PM](https://github.com/pokt-network/poktroll/assets/1892194/77c16dd6-7629-4e42-8e08-6d9e81bb1092)
![Screenshot 2024-06-05 at 1 02 35 PM](https://github.com/pokt-network/poktroll/assets/1892194/3029d092-5391-4850-84ea-d0c8ee1e9109)
![Screenshot 2024-06-05 at 1 02 39 PM](https://github.com/pokt-network/poktroll/assets/1892194/8195b47f-f79d-488d-b037-fb10be3204fe)

**Comment by Olshansk** [comment]
@rBurgett I found this thread in the cosmos docs: error while using map in txs w/ the tl;dr is that you can use maps in txs.

![Screenshot 2024-06-05 at 1 03 49 PM](https://github.com/pokt-network/poktroll/assets/1892194/2cf0219d-2a3d-48e5-9a49-df7eff865d5a)

**Comment by Olshansk** [comment]
Also, I noticed that we're "re-inventing the wheel" on how params are updated. I did so by surching for `"update-params"`

![Screenshot 2024-06-05 at 1 04 16 PM](https://github.com/pokt-network/poktroll/assets/1892194/7bc593ec-1368-4cac-9b60-cce682ed1b53)

**Comment by Olshansk** [comment]
I decided to dive deeper into the whole thing and have an alternative suggestion. CCing @bryanchriswhite and @moatus  for context too.

1. We keep the default `ComputeUnitToToken` multiple as is.
2. We update the `Service` protobuf to have a `compute_units_to_token_multiplier`

```diff
message Service {
  // For example, what if we want to request a session for a certain service but with some additional configs that identify it?
  string id = 1; // Unique identifier for the service

  // TODO_TECHDEBT: Name is currently unused but acts as a reminder that an optional onchain representation of the service is necessary
  string name = 2; // (Optional) Semantic human readable name for the service
  
  uint64 compute_units_to_token_multiplier = 3;
}
```

3. The owner of the Service (i.e. the Source owner) is the only one who can update the multiplier on their source.
4. The default for it is based on (1)
5. We won't need a global RTTM map (simplifying things)

**Comment by rBurgett** [comment]
@Olshansk that is an interesting alternative. Using a global RTTM map and serializing it as a string isn't complicated, imo, but I'll defer to your suggestion. Now, regarding number four where you say that the default should be the `compute_units_to_tokens_multiplier`, do you mean that 1) when a service is created we should set that as the service's `compute_units_to_tokens_multiplier` value? Or, 2) that should we treat the cuttm as an optional service param and fallback to the default during session settlement if a service-specific cuttm is not set?

**Comment by Olshansk** [comment]
_tl;dr Let's do the following:_

- Keep the global param variable `compute_units_to_tokens_multipler` (this will have a default value in genesis)
- Each `Service` have it's own `compute_units_per_relay` that can only be set the service owner/creator (this will need a default value upon service creation)
- Scratch the original map approach; we have found something easier

---
This is a result per an offline discussion with @moatus @RawthiL and @studna.

![Screenshot_2024-06-05_at_2 59 01_PM](https://github.com/pokt-network/poktroll/assets/1892194/efec5799-d0b4-4f34-8dc1-bdc559bdffd1)

![Screenshot 2024-06-05 at 7 47 03 PM](https://github.com/pokt-network/poktroll/assets/1892194/d05be50d-aed7-46de-95a4-170534fcb7b0)

**Comment by rBurgett** [comment]
Sounds like a plan.

**Comment by red-0ne** [comment]
Closing as per #552 and #706.
